File: lib/services/spoof_controller.dart
import 'dart:async';
import 'dart:io';

import 'package:flutter/foundation.dart';
import 'package:permission_handler/permission_handler.dart';

import '../models/lat_lng.dart';
import '../models/spoof_config.dart';
import '../utils/motion_metrics.dart';
import 'mock_location_service.dart';
import 'path_simulator.dart';
import 'secure_storage_service.dart';
import 'session_serializer.dart';

class SpoofController extends ChangeNotifier {
  SpoofController({
    MockLocationService? mockLocation,
    SecureStorageService? storage,
  })  : _mock = mockLocation ?? MockLocationService(),
        _storage = storage ?? SecureStorageService();

  final MockLocationService _mock;
  final SecureStorageService _storage;

  SpoofConfig _config = SpoofConfig();
  PathSimulator? _simulator;
  Timer? _timer;
  DateTime? _sessionStart;
  DateTime? _autoResetAt;

  bool _active = false;
  bool _paused = false;
  bool _mockAppReady = false;
  PrivacyStatus _status = PrivacyStatus.idle;
  LatLng? _lastPushed;
  LatLng? _previousPushed;
  String? _statusMessage;
  final List<LatLng> _trajectory = [];
  double _nativeProgress = 0;

  SpoofConfig get config => _config;
  bool get isActive => _active;
  bool get isPaused => _paused;
  bool get mockAppReady => _mockAppReady;
  PrivacyStatus get status => _status;
  LatLng? get currentPosition => _lastPushed ?? _simulator?.currentPosition;
  List<LatLng> get trajectory => List.unmodifiable(_trajectory);
  String? get statusMessage => _statusMessage;
  double get progress =>
      Platform.isAndroid && _active ? _nativeProgress : (_simulator?.progress ?? 0);
  Duration? get remaining {
    if (_sessionStart == null || !_active) return null;
    final total = Duration(minutes: _config.durationMinutes);
    final elapsed = DateTime.now().difference(_sessionStart!);
    final left = total - elapsed;
    return left.isNegative ? Duration.zero : left;
  }

  Future<void> loadPersisted() async {
    final saved = await _storage.loadConfig();
    if (saved != null) _config = saved;
    _mockAppReady = await _mock.isMockLocationEnabled();
    notifyListeners();
    await _recoverFromCrash();
  }

  Future<bool> checkMockReady() async {
    _mockAppReady = await _mock.isMockLocationEnabled();
    notifyListeners();
    return _mockAppReady;
  }

  Future<void> requestPermissions() async {
    await Permission.notification.request();
    await Permission.location.request();
  }

  Future<void> openDeveloperSettings() => _mock.openDeveloperSettings();

  Future<void> openAppSettings() => _mock.openAppSettings();

  Future<List<String>> readAuditLog() => _storage.readAudit();

  Future<void> _recoverFromCrash() async {
    final wasActive = await _storage.wasSessionActive();
    if (!wasActive) return;

    final ready = await _mock.initializeProvider();
    if (!ready) {
      _status = PrivacyStatus.error;
      _statusMessage = 'Recovery failed — set Smart Route as mock location app';
      notifyListeners();
      return;
    }

    if (Platform.isAndroid) {
      await _mock.startForegroundService(
        SessionSerializer.toNativeJson(_config, progress: 0),
      );
    }

    _active = true;
    _status = PrivacyStatus.fallback;
    _statusMessage =
        'Session recovered — safe zone until you tap Stop or Start';
    await _applyLocation(_config.safeZone, recordHistory: false);
    await _storage.appendAudit('Crash/boot recovery');
    notifyListeners();
  }

  Future<void> updateConfig(SpoofConfig config) async {
    _config = config;
    await _storage.saveConfig(config);
    notifyListeners();
  }

  Future<void> start() async {
    if (_active) return;

    await requestPermissions();
    _mockAppReady = await _mock.isMockLocationEnabled();
    if (!_mockAppReady) {
      _status = PrivacyStatus.error;
      _statusMessage =
          'Select Smart Route as mock location app in Developer options.';
      notifyListeners();
      return;
    }

    final ready = await _mock.initializeProvider();
    if (!ready) {
      _status = PrivacyStatus.error;
      _statusMessage = 'Could not register GPS test provider.';
      notifyListeners();
      return;
    }

    _simulator = PathSimulator(_config)..reset();
    _trajectory.clear();
    _previousPushed = null;
    _nativeProgress = 0;
    _sessionStart = DateTime.now();
    _autoResetAt = _config.autoResetMinutes != null
        ? DateTime.now().add(Duration(minutes: _config.autoResetMinutes!))
        : null;
    _active = true;
    _paused = false;
    _status = PrivacyStatus.active;
    _statusMessage =
        'Injecting mock GPS system-wide via Android test provider';

    await _storage.setSessionActive(true);
    await _storage.appendAudit('Session started (${_config.mode.name})');

    if (Platform.isAndroid) {
      await _mock.startForegroundService(
        SessionSerializer.toNativeJson(_config, progress: 0),
      );
      _startTimer(androidNative: true);
    } else {
      await _pushCurrent();
      _startTimer(androidNative: false);
    }
    notifyListeners();
  }

  Future<void> pause() async {
    if (!_active) return;
    _paused = true;
    _statusMessage = 'Paused — coordinates hold at last injection';
    await _storage.appendAudit('Paused');
    notifyListeners();
  }

  Future<void> resume() async {
    if (!_active) return;
    _paused = false;
    _statusMessage = 'Mock GPS resumed';
    await _storage.appendAudit('Resumed');
    notifyListeners();
  }

  Future<void> stop({bool userInitiated = true}) async {
    _timer?.cancel();
    _timer = null;
    _active = false;
    _paused = false;
    _simulator = null;
    _nativeProgress = 0;
    _status = PrivacyStatus.idle;
    _statusMessage = userInitiated ? 'Mock GPS stopped' : 'Session ended';

    if (Platform.isAndroid) {
      await _mock.stopForegroundService();
    }
    await _mock.removeProvider();
    await _storage.setSessionActive(false);
    await _storage.appendAudit('Stopped (user=$userInitiated)');
    notifyListeners();
  }

  void _startTimer({required bool androidNative}) {
    _timer?.cancel();
    final interval = Duration(seconds: _config.updateIntervalSeconds);
    _timer = Timer.periodic(interval, (_) {
      _tick(interval.inMilliseconds / 1000.0, androidNative: androidNative);
    });
  }

  Future<void> _tick(double deltaSeconds, {required bool androidNative}) async {
    if (!_active || _paused) return;

    if (_autoResetAt != null && DateTime.now().isAfter(_autoResetAt!)) {
      await stop(userInitiated: false);
      _statusMessage = 'Auto-reset after configured time window';
      return;
    }

    if (androidNative) {
      await _tickAndroid(deltaSeconds);
    } else {
      await _tickFlutter(deltaSeconds);
    }

    final elapsed = DateTime.now().difference(_sessionStart!);
    if (elapsed.inMinutes >= _config.durationMinutes) {
      await stop(userInitiated: false);
      _statusMessage = 'Duration limit reached';
    }
  }

  Future<void> _tickAndroid(double deltaSeconds) async {
    final status = await _mock.getSessionStatus();
    if (status != null) {
      _updateFromNative(status);
      if (status.done) {
        await stop(userInitiated: false);
        _statusMessage = 'Route complete';
      }
      return;
    }
    final advance = await _mock.advanceSession(deltaSeconds);
    if (advance == null) return;
    _updateFromNative(advance);
    if (!advance.ok) {
      _status = PrivacyStatus.fallback;
      await _applyLocation(_config.safeZone);
    }
    if (advance.done) {
      await stop(userInitiated: false);
      _statusMessage = 'Route complete';
    }
  }

  void _updateFromNative(NativeAdvanceResult r) {
    final pos = LatLng(r.lat, r.lon);
    _previousPushed = _lastPushed;
    _lastPushed = pos;
    _nativeProgress = r.progress;
    _trajectory.add(pos);
    if (_trajectory.length > 5000) _trajectory.removeAt(0);
    notifyListeners();
  }

  Future<void> _tickFlutter(double deltaSeconds) async {
    LatLng next;
    if (_config.mode == SpoofMode.staticLocation) {
      next = _config.start;
    } else {
      final sim = _simulator!;
      next = sim.tick(deltaSeconds);
      if (sim.isComplete) {
        await _applyLocation(_config.end, deltaSeconds: deltaSeconds);
        await stop(userInitiated: false);
        _statusMessage = 'Route complete';
        return;
      }
    }
    await _applyLocation(next, deltaSeconds: deltaSeconds);

    if (_lastPushed != null) {
      final leak = await _mock.detectPossibleLeak(_lastPushed!);
      if (leak) {
        _status = PrivacyStatus.leakWarning;
        _statusMessage = 'GPS divergence — safe zone applied';
        await _applyLocation(_config.safeZone, forceFallback: true);
      } else if (_status == PrivacyStatus.leakWarning) {
        _status = PrivacyStatus.active;
      }
    }
  }

  Future<void> _pushCurrent() async {
    final pos = _config.mode == SpoofMode.staticLocation
        ? _config.start
        : (_simulator?.currentPosition ?? _config.start);
    await _applyLocation(pos);
  }

  Future<void> _applyLocation(
    LatLng pos, {
    bool forceFallback = false,
    double deltaSeconds = 1,
    bool recordHistory = true,
  }) async {
    var target = pos;
    var bearing = 0.0;
    var speed = 0.0;
    if (_previousPushed != null) {
      bearing = MotionMetrics.bearingDegrees(_previousPushed!, target);
      speed = MotionMetrics.speedMps(_previousPushed!, target, deltaSeconds);
    }

    var ok = await _mock.pushLocation(
      target,
      bearing: bearing,
      speedMps: speed,
    );

    if (!ok) {
      _status = PrivacyStatus.fallback;
      target = _config.safeZone;
      ok = await _mock.pushLocation(target);
      _statusMessage = ok
          ? 'Injection failed — safe zone applied'
          : 'Mock injection failed';
      await _storage.appendAudit('Fallback to safe zone');
    } else if (!forceFallback && _status == PrivacyStatus.fallback) {
      _status = PrivacyStatus.active;
    }

    if (ok) {
      _previousPushed = _lastPushed;
      _lastPushed = target;
      if (recordHistory) {
        _trajectory.add(target);
        if (_trajectory.length > 5000) _trajectory.removeAt(0);
      }
    } else {
      _status = PrivacyStatus.error;
    }
    notifyListeners();
  }

  @override
  void dispose() {
    _timer?.cancel();
    super.dispose();
  }
}

File: app/src/main/java/com/cybersec/smartroute/service/SpoofController.java
package com.cybersec.smartroute.service;

import android.content.Context;
import android.os.Handler;
import android.os.Looper;

import androidx.annotation.MainThread;
import androidx.annotation.Nullable;

import com.cybersec.smartroute.engine.MockLocationEngine;
import com.cybersec.smartroute.engine.MockLocationPermissions;
import com.cybersec.smartroute.engine.SessionAdvancer;
import com.cybersec.smartroute.model.LatLng;
import com.cybersec.smartroute.model.PrivacyStatus;
import com.cybersec.smartroute.model.SpoofConfig;
import com.cybersec.smartroute.model.SpoofMode;
import com.cybersec.smartroute.storage.MockLocationSessionStore;
import com.cybersec.smartroute.storage.SecureStorage;
import com.cybersec.smartroute.util.LeakDetector;

import org.json.JSONObject;

import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Set;

/**
 * Singleton orchestrator: state machine, persistence, foreground-service
 * lifecycle, leak detection, and observable state for the UI.
 */
public final class SpoofController {

    public interface Listener {
        void onStateChanged(SpoofController c);
    }

    private static volatile SpoofController instance;

    public static SpoofController get(Context context) {
        if (instance == null) {
            synchronized (SpoofController.class) {
                if (instance == null) instance = new SpoofController(context.getApplicationContext());
            }
        }
        return instance;
    }

    private final Context appContext;
    private final SecureStorage storage;
    private final MockLocationSessionStore sessionStore;
    private final Handler mainHandler = new Handler(Looper.getMainLooper());
    private final Set<Listener> listeners = Collections.newSetFromMap(new java.util.concurrent.ConcurrentHashMap<>());
    private final List<LatLng> trajectory = new ArrayList<>();

    private SpoofConfig config = SpoofConfig.defaults();
    private boolean active;
    private boolean paused;
    private boolean mockAppReady;
    private PrivacyStatus status = PrivacyStatus.IDLE;
    private String statusMessage;
    private LatLng lastPushed;
    private double progress;
    private long sessionStartMs;
    private Long autoResetAtMs;
    private Runnable tick;
    private double currentSpeedMps;
    private double currentBearingDeg;
    private double distanceTraveledKm;

    private SpoofController(Context appContext) {
        this.appContext = appContext;
        this.storage = new SecureStorage(appContext);
        this.sessionStore = new MockLocationSessionStore(appContext);
        MockLocationEngine.attach(appContext);
        loadPersisted();
    }

    // ---------- listeners ----------

    public void addListener(Listener l) {
        listeners.add(l);
    }

    public void removeListener(Listener l) {
        listeners.remove(l);
    }

    private void notifyListeners() {
        if (Looper.myLooper() == Looper.getMainLooper()) {
            for (Listener l : new LinkedHashSet<>(listeners)) l.onStateChanged(this);
        } else {
            mainHandler.post(this::notifyListeners);
        }
    }

    // ---------- state ----------

    public SpoofConfig getConfig() { return config; }
    public boolean isActive() { return active; }
    public boolean isPaused() { return paused; }
    public boolean isMockAppReady() { return mockAppReady; }
    public PrivacyStatus getStatus() { return status; }
    public String getStatusMessage() { return statusMessage; }
    @Nullable public LatLng getCurrentPosition() { return lastPushed; }
    public double getProgress() { return progress; }
    public List<LatLng> getTrajectory() {
        synchronized (trajectory) {
            return Collections.unmodifiableList(new ArrayList<>(trajectory));
        }
    }
    public long getSessionStartMs() { return sessionStartMs; }
    public double getCurrentSpeedMps() { return currentSpeedMps; }
    public double getCurrentBearing() { return currentBearingDeg; }
    public double getDistanceTraveledKm() { return distanceTraveledKm; }

    public Long getRemainingMs() {
        if (!active || sessionStartMs == 0) return null;
        long total = config.durationMinutes * 60_000L;
        long left = total - (System.currentTimeMillis() - sessionStartMs);
        return Math.max(0, left);
    }

    // ---------- config ----------

    public void updateConfig(SpoofConfig newConfig) {
        this.config = newConfig;
        storage.saveConfig(newConfig);
        notifyListeners();
    }

    public List<String> readAuditLog() {
        return storage.readAudit();
    }

    public boolean checkMockReady() {
        mockAppReady = MockLocationPermissions.isMockLocationEnabled(appContext);
        notifyListeners();
        return mockAppReady;
    }

    public void openDeveloperSettings() {
        MockLocationPermissions.openDeveloperSettings(appContext);
    }

    public void openAppSettings() {
        MockLocationPermissions.openAppSettings(appContext);
    }

    /** Force UI sync from the active session store (e.g. after returning from background). */
    public void refreshUiFromSession() {
        if (!active || config.mode == SpoofMode.STATIC_LOCATION) {
            notifyListeners();
            return;
        }
        syncUiFromSession();
    }

    private void loadPersisted() {
        SpoofConfig saved = storage.loadConfig();
        if (saved != null) config = saved;
        mockAppReady = MockLocationPermissions.isMockLocationEnabled(appContext);
        recoverFromCrashIfNeeded();
        notifyListeners();
    }

    private void recoverFromCrashIfNeeded() {
        if (!storage.wasSessionActive()) return;
        if (!sessionStore.isSessionActive()) {
            storage.setSessionActive(false);
            return;
        }
        JSONObject session = sessionStore.loadSession();
        if (session == null) {
            storage.setSessionActive(false);
            return;
        }
        if (!MockLocationEngine.initProvider()) {
            status = PrivacyStatus.ERROR;
            statusMessage = "Recovery failed — set Smart Route as the mock location app";
            storage.setSessionActive(false);
            sessionStore.clearSession();
            return;
        }
        active = true;
        paused = false;
        sessionStartMs = System.currentTimeMillis();
        status = PrivacyStatus.FALLBACK;
        statusMessage = "Session recovered — resuming mock GPS";
        MockLocationForegroundService.start(appContext, session.toString());
        startTimer();
        storage.appendAudit("Crash/boot recovery");
    }

    // ---------- lifecycle ----------

    @MainThread
    public boolean start() {
        if (active) return true;
        mockAppReady = MockLocationPermissions.isMockLocationEnabled(appContext);
        if (!mockAppReady) {
            status = PrivacyStatus.ERROR;
            statusMessage = "Select Smart Route as mock location app in Developer options.";
            notifyListeners();
            return false;
        }
        if (!MockLocationEngine.initProvider()) {
            status = PrivacyStatus.ERROR;
            statusMessage = "Could not register GPS test provider.";
            notifyListeners();
            return false;
        }

        synchronized (trajectory) { trajectory.clear(); }
        progress = 0;
        lastPushed = null;
        currentSpeedMps = 0;
        currentBearingDeg = 0;
        distanceTraveledKm = 0;
        sessionStartMs = System.currentTimeMillis();
        autoResetAtMs = config.autoResetMinutes == null
                ? null
                : sessionStartMs + config.autoResetMinutes * 60_000L;
        active = true;
        paused = false;
        status = PrivacyStatus.ACTIVE;
        statusMessage = "Injecting mock GPS via Android test provider";
        storage.setSessionActive(true);
        storage.appendAudit("Session started (" + config.mode.wireName() + ")");

        SessionAdvancer.resetDebounce();
        JSONObject engineSession = config.toEngineJson(0);
        sessionStore.saveSession(engineSession);
        MockLocationForegroundService.start(appContext, engineSession.toString());

        startTimer();
        syncUiFromSession();
        notifyListeners();
        return true;
    }

    @MainThread
    public void pause() {
        if (!active) return;
        paused = true;
        statusMessage = "Paused — coordinates hold at last injection";
        storage.appendAudit("Paused");
        notifyListeners();
    }

    @MainThread
    public void resume() {
        if (!active) return;
        paused = false;
        statusMessage = "Mock GPS resumed";
        storage.appendAudit("Resumed");
        notifyListeners();
    }

    @MainThread
    public void stop(boolean userInitiated) {
        if (tick != null) {
            mainHandler.removeCallbacks(tick);
            tick = null;
        }
        active = false;
        paused = false;
        progress = 0;
        currentSpeedMps = 0;
        currentBearingDeg = 0;
        status = PrivacyStatus.IDLE;
        statusMessage = userInitiated ? "Mock GPS stopped" : "Session ended";
        MockLocationForegroundService.stop(appContext);
        sessionStore.clearSession();
        MockLocationEngine.removeProvider();
        storage.setSessionActive(false);
        storage.appendAudit("Stopped (user=" + userInitiated + ")");
        notifyListeners();
    }

    // ---------- internal ticking ----------

    private void startTimer() {
        if (tick != null) mainHandler.removeCallbacks(tick);
        long intervalMs = config.updateIntervalSeconds * 1000L;
        tick = new Runnable() {
            @Override
            public void run() {
                if (!active) return;
                if (!paused) advance(intervalMs / 1000.0);
                if (active) mainHandler.postDelayed(this, intervalMs);
            }
        };
        mainHandler.postDelayed(tick, intervalMs);
    }

    private void advance(double deltaSeconds) {
        if (autoResetAtMs != null && System.currentTimeMillis() > autoResetAtMs) {
            stop(false);
            statusMessage = "Auto-reset after configured window";
            notifyListeners();
            return;
        }
        if (config.mode == SpoofMode.STATIC_LOCATION) {
            applyLocation(config.start, false, deltaSeconds, true);
            checkLeak();
            return;
        }

        SessionAdvancer.AdvanceResult r = SessionAdvancer.pollUiState(appContext);
        if (r == null) {
            if (active) stop(false);
            return;
        }
        applyAdvanceResult(r, deltaSeconds);
        checkLeak();
    }

    private void syncUiFromSession() {
        SessionAdvancer.AdvanceResult r = SessionAdvancer.pollUiState(appContext);
        if (r == null) {
            mainHandler.postDelayed(this::syncUiFromSession, 150);
            return;
        }
        applyAdvanceResult(r, 0.0);
    }

    private void applyAdvanceResult(SessionAdvancer.AdvanceResult r, double deltaSeconds) {
        if (r == null) return;
        LatLng pos = new LatLng(r.lat, r.lon);
        boolean moved = lastPushed == null || LatLng.haversineKm(lastPushed, pos) > 0.00001;
        if (moved && lastPushed != null) {
            distanceTraveledKm += LatLng.haversineKm(lastPushed, pos);
        }
        currentSpeedMps = r.speedMps;
        currentBearingDeg = r.bearing;
        progress = r.progress;
        lastPushed = pos;
        if (moved) {
            synchronized (trajectory) {
                trajectory.add(pos);
                if (trajectory.size() > 5000) trajectory.remove(0);
            }
        }
        if (!r.ok) {
            status = PrivacyStatus.FALLBACK;
            applyLocation(config.safeZone, true, deltaSeconds, false);
        } else if (status == PrivacyStatus.FALLBACK) {
            status = PrivacyStatus.ACTIVE;
        }
        if (System.currentTimeMillis() - sessionStartMs >= config.durationMinutes * 60_000L) {
            stop(false);
            statusMessage = "Duration limit reached";
            notifyListeners();
            return;
        }
        if (r.done) {
            stop(false);
            statusMessage = "Route complete";
            notifyListeners();
            return;
        }
        notifyListeners();
    }

    private void applyLocation(LatLng pos, boolean fallback, double dt, boolean recordHistory) {
        float bearing = 0f;
        float speed = 0f;
        if (lastPushed != null) {
            bearing = (float) com.cybersec.smartroute.util.MotionMetrics.bearingDegrees(lastPushed, pos);
            speed = (float) com.cybersec.smartroute.util.MotionMetrics.speedMps(lastPushed, pos, dt);
            distanceTraveledKm += LatLng.haversineKm(lastPushed, pos);
        }
        currentBearingDeg = bearing;
        currentSpeedMps = speed;
        boolean ok = MockLocationEngine.setLocation(
                pos.latitude, pos.longitude, 3f, 0d, bearing, speed);
        if (!ok && !fallback) {
            status = PrivacyStatus.FALLBACK;
            ok = MockLocationEngine.setLocation(
                    config.safeZone.latitude, config.safeZone.longitude);
            statusMessage = ok ? "Injection failed — safe zone applied" : "Mock injection failed";
            storage.appendAudit("Fallback to safe zone");
            pos = config.safeZone;
        }
        if (ok) {
            lastPushed = pos;
            if (recordHistory) {
                synchronized (trajectory) {
                    trajectory.add(pos);
                    if (trajectory.size() > 5000) trajectory.remove(0);
                }
            }
            try {
                JSONObject session = sessionStore.loadSession();
                if (session != null) {
                    session.put("lastLat", pos.latitude);
                    session.put("lastLon", pos.longitude);
                    sessionStore.saveSession(session);
                }
            } catch (Exception ignored) {
            }
        } else {
            status = PrivacyStatus.ERROR;
        }
        notifyListeners();
    }

    private void checkLeak() {
        final LatLng injected = lastPushed;
        if (injected == null) return;
        LeakDetector.check(appContext, injected, 150.0, leak -> {
            if (leak) {
                status = PrivacyStatus.LEAK_WARNING;
                statusMessage = "GPS divergence — safe zone applied";
                storage.appendAudit("Leak detected — fallback");
                applyLocation(config.safeZone, true, 1.0, false);
            } else if (status == PrivacyStatus.LEAK_WARNING) {
                status = PrivacyStatus.ACTIVE;
                notifyListeners();
            }
        });
    }
}
